src/main/java/com/scalar/cassy/transferer/AzureBlobFileDownloader.java

```diff
@@ -1,63 +1,102 @@
+import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import reactor.core.publisher.Mono;
 
 public class AzureBlobFileDownloader implements FileDownloader {
   private static final Logger logger = LoggerFactory.getLogger(AzureBlobFileDownloader.class);
-  private BlobContainerAsyncClient blobContainerClient;
+  private BlobContainerClient blobContainerClient;
+  private static final int NUM_THREADS = 3;
+  private static final int ASYNC_FILE_DOWNLOAD_LIMIT = 20;
+  private final ExecutorService executorService = Executors.newFixedThreadPool(NUM_THREADS);
 
   @Inject
-  public AzureBlobFileDownloader(BlobContainerAsyncClient blobContainerClient) {
+  public AzureBlobFileDownloader(BlobContainerClient blobContainerClient) {
     this.blobContainerClient = blobContainerClient;
   }
 
   @Override
   public void download(RestoreConfig config) {
+    List<Future<Void>> downloadFuture = new ArrayList<>();
     String key = BackupPath.create(config, config.getKeyspace());
 
     logger.info("Downloading " + blobContainerClient.getBlobContainerName() + "/" + key);
-    List<Mono<BlobProperties>> filesToBeDownloaded = new ArrayList<>();
-    Iterable<BlobItem> keyspaceBlobs =
-        blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/")).toIterable();
-    for (BlobItem blob : keyspaceBlobs) {
+    for (BlobItem blob : listBlobs(key)) {
       Path destFile = Paths.get(config.getDataDir(), blob.getName());
       try {
         Files.createDirectories(destFile.getParent());
       } catch (IOException e) {
         throw new FileTransferException(e);
       }
-      filesToBeDownloaded.add(
-          blobContainerClient
-              .getBlobAsyncClient(blob.getName())
-              .downloadToFile(destFile.toString())
-              .doOnSuccess(
-                  blobProperties -> logger.info("Download file succeeded : " + destFile.toString()))
-              .doOnError(
-                  error -> {
-                    throw new FileTransferException(
-                        "Download file failed : " + destFile.toString(), error);
-                  }));
+      downloadFuture.add(
+          executorService.submit(
+              () -> {
+                try (OutputStream outputStream = writeStream(destFile)) {
+                  blobContainerClient.getBlobClient(blob.getName()).download(outputStream);
+
+                } catch (IOException e) {
+                  throw new FileTransferException(e);
+                }
+                logger.info("Download file succeeded : " + destFile.toString());
```

Comment:
```suggestion
                logger.info("Download file succeeded : " + destFile);
```

Reply:
Sorry for the inconvenience.
Suggested change integrated 328ca47